internal/controller/postgrescluster/util.go

```diff
@@ -285,3 +290,188 @@ func safeHash32(content func(w io.Writer) error) (string, error) {
+
+	// Capture the largest volume size currently defined for a given instance set.
+	// This value will capture our desired update.
+	volumeRequestSize := pvc.Spec.Resources.Requests.Storage()
+
+	// If the request value is greater than the set limit, use the limit and issue
+	// a warning event. A limit of 0 is ignorned.
+	if !volumeLimitFromSpec.IsZero() &&
+		volumeRequestSize.Value() > volumeLimitFromSpec.Value() {
+		r.Recorder.Eventf(cluster, corev1.EventTypeWarning, "VolumeRequestOverLimit",
+			"%s volume request (%v) for %s/%s is greater than set limit (%v). Limit value will be used.",
+			volumeType, volumeRequestSize, cluster.Name, instanceSpecName, volumeLimitFromSpec)
+
+		pvc.Spec.Resources.Requests = corev1.ResourceList{
+			corev1.ResourceStorage: *resource.NewQuantity(volumeLimitFromSpec.Value(), resource.BinarySI),
+		}
+		// Otherwise, if the limit is not set or the feature gate is not enabled, do not autogrow.
+	} else if !volumeLimitFromSpec.IsZero() && feature.Enabled(ctx, feature.AutoGrowVolumes) {
+
+		// determine the appropriate volume request based on what's set in the status
+		if dpv, err := getDesiredVolumeSize(
+			cluster, volumeType, instanceSpecName, volumeRequestSize,
+		); err != nil {
+			log.Error(err, "For "+cluster.Name+"/"+instanceSpecName+
+				": Unable to parse "+volumeType+" volume request: "+dpv)
+		}
+
+		// If the volume request size is greater than or equal to the limit and the
+		// limit is not zero, update the request size to the limit value.
+		// If the user manually requests a lower limit that is smaller than the current
+		// or requested volume size, it will be ignored in favor of the limit value.
+		if volumeRequestSize.Value() >= volumeLimitFromSpec.Value() {
+
+			r.Recorder.Eventf(cluster, corev1.EventTypeNormal, "VolumeLimitReached",
+				"%s volume(s) for %s/%s are at size limit (%v).", volumeType,
+				cluster.Name, instanceSpecName, volumeLimitFromSpec)
+
+			// If the volume size request is greater than the limit, issue an
+			// additional event warning.
+			if volumeRequestSize.Value() > volumeLimitFromSpec.Value() {
+				r.Recorder.Eventf(cluster, corev1.EventTypeWarning, "DesiredVolumeAboveLimit",
+					"The desired size (%v) for the %s/%s %s volume(s) is greater than the size limit (%v).",
+					volumeRequestSize, cluster.Name, instanceSpecName, volumeType, volumeLimitFromSpec)
+			}
+
+			volumeRequestSize = volumeLimitFromSpec
+		}
+		pvc.Spec.Resources.Requests = corev1.ResourceList{
+			corev1.ResourceStorage: *resource.NewQuantity(volumeRequestSize.Value(), resource.BinarySI),
+		}
+	}
+}
+
+// getDesiredVolumeSize compares the volume request size to the suggested autogrow
+// size stored in the status and updates the value when the status value is larger.
+func getDesiredVolumeSize(cluster *v1beta1.PostgresCluster,
+	volumeType, instanceSpecName string,
+	volumeRequestSize *resource.Quantity) (string, error) {
+
+	switch volumeType {
```

Comment:
:memo: This switch statement will likely be very similar to the one used by `limitIsSet`. The difference is that this function references status objects that don't yet exist whereas `limitIsSet` required nothing new outside of itself, so it made sense to move forward a bit there but not here.